webgpu: Update wgpu and revamp computepass

sagudev opened the pull request:
wgpu's compute pass object now lives only in wgpu thread, functions return compute pass as object instead of an id, so we need to do id mapping ourself.


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes in WebGPU CTS

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

sagudev committed:
Do not wait on drop, but rather wake poller thread

sagudev committed:
Update wgpu and render stuff

sagudev committed:
Set some good expectations

sagudev committed:
Update wgpu again

sagudev commented:
try run: https://github.com/sagudev/servo/actions/runs/9602065915/job/26483847356

> Unexpected validation error occurred: At least one color attachment or depth-stencil attachment was expected, but no render target for the pipeline was specified.

is actually expected per spec: https://github.com/gfx-rs/wgpu/pull/5715

sagudev committed:
handle IPC error as warning

sagudev committed:
More good expectations

mrobinson commented on Cargo.toml:
```diff
@@ -135,8 +135,8 @@ webpki-roots = "0.25"
 webrender_traits = { path = "components/shared/webrender" }
-wgpu-core = { git = "https://github.com/gfx-rs/wgpu", rev = "d0a5e48aa7e84683114c3870051cc414ae92ac03" }
-wgpu-types = { git = "https://github.com/gfx-rs/wgpu", rev = "d0a5e48aa7e84683114c3870051cc414ae92ac03" }
+wgpu-core = { git = "https://github.com/gfx-rs/wgpu", rev = "1e784c9c0af7a785b62e3b6840ed012a7477520f" }
+wgpu-types = { git = "https://github.com/gfx-rs/wgpu", rev = "1e784c9c0af7a785b62e3b6840ed012a7477520f" }
```
Is it possible to switch to a release of wgpu instead of using a git hash?

sagudev replied:
No, unfortunately not. Releases are cut too slowly and last ones had some nasty bugs. Also I do some work on wgpu and keeping servo up to date makes it easier to squash bugs in servo and wgpu at the same time.

BTW firefox also uses git rev instead of releases updated weekly, for similar reasons.

sagudev committed:
Some more expectations

CTS does not match the spec: https://github.com/gpuweb/cts/issues/3806

sagudev committed:
This expectations are due to other changes in servo

also happening on main

sagudev committed:
Explain error_command_encoders and remove RefCell around it

sagudev commented:
More work on command buffers/encoders is left for a followup PR.

gterzian commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -218,7 +223,9 @@ impl WGPU {
                         };
 
+                        // invalidate command buffer too
                         self.encoder_record_error(command_encoder_id, &result);
```
what do you mean? Also a nit for another time: that method should also be called `maybe_`.

sagudev replied:
Specs says that we need to invalidate command encoder on error, but wgpu does not do this (yet). So we consider command encoder invalid if there is error stored in hashmap of error encoders. When command encoder is finished into command buffer (wgpu uses same id via id transmutes) we need to make sure that invalid command encoder also finishes into invalid command buffer (or if there is error when doing finish) so we reuse same error hashmap for this.

gterzian commented on components/script/dom/gpudevice.rs:
```diff
@@ -769,6 +771,7 @@ impl GPUDeviceMethods for GPUDevice {
             Some(wgpu_pipe::RenderPipelineDescriptor {
                 label: convert_label(&descriptor.parent.parent),
                 layout,
+                cache: None,
```
So the idea is this will move to the backend thread as well right?

sagudev replied:
Yes, when wgpu changes how renderpasses work.

gterzian reviewed (changes requested):
Couple of things...

sagudev commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -73,13 +73,19 @@ pub(crate) struct WGPU {
     poller: Poller,
+    // We need to store passes here
+    compute_passes: HashMap<ComputePassId, Box<dyn DynComputePass>>,
+    //render_passes: HashMap<RenderPassId, Box<dyn DynRenderPass>>,
```
Wgpu will also change how renderpass works, so we will need this in near future.

gterzian commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -745,35 +751,87 @@ impl WGPU {
-                        self.encoder_record_error(command_encoder_id, &result);
+                            self.dispatch_error(
+                                device_id,
+                                Error::Validation("Render pass already ended".to_string()),
```
Well, I guess `!valid` means ended, actually...

sagudev replied:
Actually, this validation in script (content time-line) should actually be done in webgpu thread (usually it's done within wgpu). That's actually the gist of what needs to be done in servo about webgpu, we have a lot of logic in script that should be in device timeline (webgpu thread), usually already in wgpu. Maybe the specs were different when the code was written.

Here I did some ground work for changes of render pass that will come (mostly because some changes around it are already included in wgpu), but I plan to do complete refresh of RenderPass (like I did in for compute pass here), when all necessary changes land in wgpu.

sagudev commented on components/script/dom/gpucomputepassencoder.rs:
```diff
@@ -79,57 +77,79 @@ impl GPUComputePassEncoderMethods for GPUComputePassEncoder {
-            GPUCommandEncoderState::Open,
-            GPUCommandEncoderState::EncodingComputePass,
-        );
+            .expect("Failed to send EndComputePass"); //TODO: handle error
```
Because is just refactored old code, but I will fix it.

sagudev commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -745,35 +751,87 @@ impl WGPU {
                                 command_encoder_id,
-                                &pass
-                            )).map_err(|e| format!("{:?}", e))
+                                &ComputePassDescriptor { label: None, timestamp_writes: None }
```
Generally we just pass wgpu descriptors from script, but `ComputePassDescriptor` is not serializable.

sagudev commented on components/script/dom/gpucommandencoder.rs:
```diff
@@ -122,32 +105,24 @@ impl GPUCommandEncoderMethods for GPUCommandEncoder {
-        };
+        if let Err(e) = self.channel.0.send(WebGPURequest::BeginComputePass {
+            command_encoder_id: self.id().0,
+            compute_pass_id,
```
I will only put label field as we do not support timestamps yet.

sagudev commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -745,35 +751,87 @@ impl WGPU {
+                        ));
+                        self.compute_passes.insert(compute_pass_id, pass);
+                    },
+                    WebGPURequest::ComputePassSetPipeline {
```
When we end pass we must drop it, or else it will still hold resources (and queue_submit will panic).

sagudev committed:
fixup

sagudev commented:
Hm, some validation still fails, apparently command encoders are not invalidated on invalid passes ending.

sagudev committed:
store validness of passes

sagudev committed:
More good expectations

gterzian commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -84,7 +84,7 @@ pub(crate) struct WGPU {
     poller: Poller,
     // We need to store passes here
-    compute_passes: HashMap<ComputePassId, Box<dyn DynComputePass>>,
+    compute_passes: HashMap<ComputePassId, (Box<dyn DynComputePass>, bool)>,
```
A doc for `valid` would be good.

sagudev replied:
Done!

gterzian reviewed (approved):
LGTM, with one nit

sagudev committed:
More docs

sagudev committed:
this assert is wrong

sagudev committed:
This is even more right per CTS/spec

Only Command encoder state errors are allowed here, but wgpu does not exposes them.

sagudev committed:
More good expectations

sagudev committed:
One bad expectation

sagudev commented:
Currently `webgpu:api,validation,encoding,encoder_state:call_after_successful_finish:*` fails due to different encoder state handling in wgpu, we could probably write more logic to make it work, but I would rather work on fixing wgpu itself.

gterzian commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -74,16 +74,17 @@ pub(crate) struct WGPU {
+    /// Provides access to poller thread
     poller: Poller,
-    // We need to store passes here
+    /// Store compute passes (that have not ended yet) and their validness
```
I think `validity` is the right word.

gterzian commented on components/webgpu/wgpu_thread.rs:
```diff
@@ -1165,10 +1166,8 @@ impl WGPU {
-                            self.compute_passes.remove(&id).is_some(),
-                            "ComputePass should exist"
-                        );
+                        // Compute pass might already ended
```
needs a "have".

gterzian reviewed (approved):
LGMT, few nits again.

sagudev committed:
Fix my english
